File: ElencoPersonePython/UserInterface.py
from os import system, name
from ListOfPersone import Elenco
import sys
from Persona import Persona
from singleton import singleton
from datetime import datetime

@singleton
class UserInterface:

    def __init__(self):
        self.__menu = ['1. Inserisci', '2. Cancella', '3. Cerca', '4. Modifica', '5 .Visualizza Elenco', '6 .Esci']
        
    
    def clear(self):
        if name == 'nt':
            _ = system('cls')
        else:
            _ = system('clear')

    def choice(self, c : int):

        while(c < 1 or c > 6):
            print('>> ')
            c = int(input())
        return c
    
    def callMenu(self, elenco: Elenco):
        while True:
            self.clear()
            print("\n".join(self.__menu))
            try:
                c = int(input("\nSeleziona un'opzione: "))
                if c == 1:
                    self.Inserisci(elenco)
                elif c == 2:
                    self.Cancella(elenco)
                elif c == 3:
                    self.Cerca(elenco)
                elif c == 4:
                    self.Modifica(elenco)
                elif c == 5:
                    self.Visualizza(elenco)
                elif c == 6:
                    print("Uscita...")
                    sys.exit()
                else:
                    print("Opzione non valida!")
            except ValueError:
                print("Inserisci un numero valido.")
            input("\nPremi Invio per continuare...")


    def Inserisci(self, elenco: Elenco):    
        nome = input('Inserisci nome: ')
        while True:
            DoB = input('Inserisci Data di Nascita (DD-MM-YYYY): ')
            try:
                datetime.strptime(DoB, '%d-%m-%Y')
                break
            except ValueError:
                print("Data non valida. Riprova.")

        p = Persona(nome, DoB)
        elenco.add(p)
        print("Persona aggiunta con successo.")

    def Cancella(self, elenco: Elenco):
        persona = self.Cerca(elenco)
        elenco.delede(persona)
        print("Persona cancellata con successo.")

    def Cerca(self, elenco: Elenco):
        print('Inserisci nome: ')
        nome = input()
        while True:
            DoB = input('Inserisci Data di Nascita (DD-MM-YYYY): ')
            try:
                datetime.strptime(DoB, '%d-%m-%Y')
                break
            except ValueError:
                print("Data non valida. Riprova.")

        if elenco.search(Persona(nome, DoB)) != None:
            print('Persona trovata')
            #print('Trovato a posizone:  ' + str(elenco.search(Persona(nome, DoB))) )
            return Persona(nome, DoB)
        else:
            print('Persona non trovata')
            return None
    
    def Modifica(self, elenco: Elenco): 
        nome = input('Inserisci nome da modificare: ')
        while True:
            DoB = input('Inserisci Data di Nascita (DD-MM-YYYY): ')
            try:
                datetime.strptime(DoB, '%d-%m-%Y')
                break
            except ValueError:
                print("Data non valida. Riprova.")
        persona = Persona(nome, DoB)
        elenco.search(persona)
        
        if persona:
            print(f"Modificando: {persona}")
            elenco.delede(persona)
            nuovo_nome = input('Inserisci nuovo nome: ')
            nuovo_DoB = input('Inserisci nuova Data di Nascita (DD-MM-YYYY): ')
            persona.nome = nuovo_nome
            persona.DoB = nuovo_DoB
            elenco.add(persona)
            print("Persona modificata con successo.")
        else:
            print("Persona non trovata.")


    def Visualizza(self, elenco: Elenco):
        print(elenco)

if __name__ == '__main__':
    print('Errore, questo file non è eseguibile')

File: ElencoPersoneJava/UserInterface.java
package ElencoPersone;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Scanner;

public class UserInterface {
    public static UserInterface _instance = null;
    private int id;

    private static final BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

    private UserInterface() {
        this.id = 0;
    }

    public static UserInterface getInstance() {
        if (UserInterface._instance == null) {
            UserInterface._instance = new UserInterface();
        }
        return UserInterface._instance;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    /*
    *
    * Fuzioni del Menu
    *
    * */

    public static void clearConsole() {
        try {
            final String os = System.getProperty("os.name").toLowerCase();
            if (os.contains("win")) {
                new ProcessBuilder("cmd", "/c", "cls").inheritIO().start().waitFor();
            } else {
                System.out.println("\033[H\033[2J");
                System.out.flush();
            }
        } catch (Exception e) {
            System.out.println("Error clearing the console.");
        }
    }

    public int showMenu() {
        Scanner scanner = new Scanner(System.in);
        System.out.println("\n=== Menu ===");
        System.out.println("1. Add person");
        System.out.println("2. Delete person");
        System.out.println("3. Show list of people");
        System.out.println("4. Exit");
        System.out.print("Seleziona un'opzione: ");

        int choice;
        while (true) {
            try {
                choice = Integer.parseInt(scanner.nextLine());
                if (choice >= 1 && choice <= 4) break;
                System.out.println("Opzione non valida. Riprova.");
            } catch (NumberFormatException e) {
                System.out.println("Input non valido. Inserisci un numero.");
            }
        }
        return choice;
    }

    public void callMenu(ListOfPersone manager) {
        while (true) {
            int choice = showMenu();
            switch (choice) {
                case 1 -> AddPerson_UserInterface(manager);
                case 2 -> DelPerson_UserInterface(manager);
                case 3 -> PrintLoP(manager);
                case 4 -> {
                    System.out.println("Uscita dal programma.");
                    return;
                }
                default -> System.out.println("Errore: scelta non valida.");
            }
        }
    }

    /*
     *
     *  Funzioni di UserInterface
     *
     * */
    private static void AddPerson_UserInterface(ListOfPersone manager) {
        try {
            System.out.print("Inserisci un nome: ");
            String nome = reader.readLine();

            System.out.print("Inserisci un cognome: ");
            String cognome = reader.readLine();

            String DoB = Utilities.readValidDate();

            clearConsole();
            Persona p = new Persona(nome, cognome, DoB);
            if (manager.Add(p)) {
                System.out.println("Persona aggiunta con successo!");
            } else {
                System.out.println("Errore: impossibile aggiungere la persona.");
            }
        } catch (IOException e) {
            System.out.println("Errore durante l'inserimento dei dati: ");
        }
    }

    private static void DelPerson_UserInterface(ListOfPersone manager) {
        try {
            System.out.print("Inserisci il nome della persona da eliminare: ");
            String nome = reader.readLine();

            System.out.print("Inserisci il cognome della persona da eliminare: ");
            String cognome = reader.readLine();

            System.out.print("Inserisci la data di nascita (YYYY-MM-DD) della persona da eliminare: ");
            String DoB = reader.readLine();

            clearConsole();
            Persona p = new Persona(nome, cognome, DoB);
            if (manager.Del(p)) {
                System.out.println("Persona eliminata con successo!");
            } else {
                System.out.println("Errore: persona non trovata.");
            }
        } catch (IOException e) {
            System.out.println("Errore durante l'inserimento dei dati: ");
        }
    }

    private static void PrintLoP(ListOfPersone manager) {
        clearConsole();
        System.out.println("\n=== Lista delle Persone ===");
        if (manager.isEmpty()) {
            System.out.println("Nessuna persona presente.");
            return;
        }

        for (Persona persona : manager.getAll()) {
            System.out.println(persona.toString());
        }
    }
}


